main/src/mill/modules/Assembly.scala

```diff
@@ -50,26 +69,23 @@ object Assembly {
       case r @ Rule.Exclude(path) => path -> r
     }.toMap
 
-    val appendPatterns = assemblyRules.collect {
-      case Rule.AppendPattern(pattern) => pattern.asPredicate().test(_)
-    }
-
-    val excludePatterns = assemblyRules.collect {
-      case Rule.ExcludePattern(pattern) => pattern.asPredicate().test(_)
+    val matchPatterns = assemblyRules.collect {
+      case r : Rule.AppendPattern => r.pattern.asPredicate() -> r
+      case r @ Rule.ExcludePattern(pattern) => pattern.asPredicate() -> r
     }
 
     mappings.foldLeft(Map.empty[String, GroupedEntry]) {
       case (entries, (mapping, entry)) =>
-        rulesMap.get(mapping) match {
-          case Some(_: Assembly.Rule.Exclude) =>
+        rulesMap.get(mapping) orElse matchPatterns.find(_._1.test(mapping)).map(_._2) match {
```

Comment:
You could split the two parts into two properly named variables and then do `orElse` separately.

Reply:
Want to notice that argument in `scala.Option#orElse` is "by-name", so extracting variables will degrade performance, or did you want me to extract functions?

Reply:
Then make the parameter a `def` so the result will be equivalent to a by-name parameter.